extension/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+
+use std::{
+    collections::VecDeque,
+    mem::size_of,
+};
+
+use timescale_extension_utils::{
+    datum::ToDatum,
+    elog,
+    pg_agg,
+    pg_sys::{
+        ArrayType,
+        Datum,
+        FLOAT8OID,
+        FLOAT8PASSBYVAL,
+        TimestampTz,
+        construct_md_array,
+    },
+};
+
+type Milliseconds = i64;
+const USECS_PER_SEC: i64 = 1_000_000;
+const USECS_PER_MS: i64 = 1_000;
+
+pg_agg!{
+    // prom divides time into no-sliding windows of fixed size, e.g.
+    // |  5 seconds  |  5 seconds  |  5 seconds  |  5 seconds  |  5 seconds  |
+    // we take the first and last values in that bucket and uses `last-first` as the
+    // value for that bucket.
+    //  | a b c d e | f g h i | j   k |   m    |
+    //  |   e - a   |  i - f  | k - j | <null> |
+    pub fn gapfill_delta_transition(
+        state: Option<Pox<GapfillDeltaTransition>>,
+        lowest_time: TimestampTz,
```

Comment:
so we have to discuss the api we want. In prom start is before the the - step and here after. I know we discussed this before but let's reconsider after talking more broadly about the sql api